refactor (crc/machine) : Provide a dummy implementation for virtualMachine object for writing unit tests (#4407)

rohanKanojia opened the pull request:
## Description


**Fixes:** Issue #4407

**Relates to:** Issue #4407, PR https://github.com/crc-org/crc/pull/4400

## Type of change
<!---
What types of changes does your code introduce? Put an `x` in all the boxes that apply
-->
 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [ ] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [x] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [x] I have read the [contributing guidelines](https://github.com/crc-org/crc/blob/main/developing.adoc)
 - [x] My code follows the style guidelines of this project
 - [x] I Keep It Small and Simple: The smaller the PR is, the easier it is to review and have it merged
 - [x] I use [conventional commits](https://www.conventionalcommits.org/) in my commit messages
 - [x] I have performed a self-review of my code
 - [x] I have added tests that prove my fix is effective or that my feature works
 - [x] New and existing unit tests pass locally with my changes
 - [x] I tested my code on specified platforms
   - [x] Linux
   - [ ] Windows
   - [ ] MacOS

## Solution/Idea

+ Add a VirtualMachine interface and make the CRC `machine` package client use the VirtualMachine interface instead of a concrete implementation. This way we can inject a dummy test FakeVirtualMachine implementation into client tests that can ease writing tests for this package.
  - Add some additional methods in VirtualMachine interface so that we can replace direct usage of struct fields with interface methods
    - `Bundle()`
    - `Driver()`
    - `API()`
    - `GetHost()`
+ Add a new field `VirtualMachine` in the client so we can avoid creating it if it's already created.
+ Introduce a new constructor method `newClientWithVirtualMachine` in machine client that would have an additional VirtualMachine argument, this would be kept package private so that it's used only by tests in the same package.
+ Add FakeVirtualMachine sturct in `fakemachine` for adding dummy implementation for `VirtualMachine` interface. Currently, I've only completed methods used by `stop_test.go`, I'll add more in small increments as I get more familiar with the project. 



## Proposed changes
- Add new `VirtualMachine` interface and make client member functions use it instead of `virtualMachine` struct.
- Add new constructor in machine client to be able to override `VirtualMachine` implementation from tests
- These files have been modified to use `VirtualMachine` interface instead of `virtualMachine` struct (these changes are only related to changing use of `virtualMachine` struct to `VirtualMachine` interface:
  - `pkg/crc/machine/ip.go`
  - `pkg/crc/machine/poweroff.go`
  - `pkg/crc/machine/start.go`
  - `pkg/crc/machine/status.go`
  - `pkg/crc/machine/stop.go`


## Testing

It's a small refactor. I've only run E2E tests locally to verify if these changes don't introduce any kind of regression.

rohanKanojia committed:
test (machine) : add vsock interaction methods to VirtualMachine interface

Make VirtualMachine implement vsock methods so that vsock interaction
is also done via VirtualMachine interface. This would help in writing
tests, we can override the behavior of expose/unexpose in fake vm
implementation.

Signed-off-by: Rohan Kumar <rohaan@redhat.com>

cfergeau commented on pkg/crc/machine/client.go:
```diff
@@ -30,21 +30,30 @@ type Client interface {
+		config:         config,
+		diskDetails:    memoize.NewMemoizer(time.Minute, 5*time.Minute),
+		ramDetails:     memoize.NewMemoizer(30*time.Second, 2*time.Minute),
+		virtualMachine: vm,
```
I'd separate the commit adding this `virtualMachine: vm` field from the commit introducing the `VirtualMachine` interface. This commit is huge because of the introduction of this interface, but the changes needed for the interface changes are very mechanical (`vm.bundle` -> `vm.Bundle`) and such. If they are put in a commit doing only this, it's easier to review as we can focus on one thing, these mechanical changes.
Then in a 2nd commit come the "logic" changes which add a way to mock/test the virtual machine, and it's going to be a lot smaller, and easier to understand/review.

rohanKanojia replied:
I have split this PR into multiple commits, please review if it's as per your expectations. I can adjust if additional changes are needed.

cfergeau replied:
Hmm https://github.com/cfergeau/crc/commits/pr/issue4407/ is closer to what I had in mind. One commit doing only the VirtualMachine interface introduction and making use of its methods (could be 2 commits if you prefer, but I think it's fine as one). I added 2 follow-up commits which are meant to be squashed in the first one.
Then one commit adds `newClientWithVirtualMachine`, `loadVirtualMachineLazily` so that we can override it, and the rest of your series, I haven't touched it.

cfergeau replied:
By the way, the content in `Solutions/Idea` and `Proposed Changes` would be nice to have in the individual commit logs, as one line commit logs are very terse when you'll get back to them in 6 months trying to remember what the commit was about, but also too long for a short log, they overflow quite a bit in github UI, git log, ... They typically should be about 60 characters in length.

Better to have a short 1line log, followed by more details in the log explaining why you are making the change, and giving a high level overview of what you are doing if it's a big change.

cfergeau commented on pkg/crc/machine/vsock.go:
```diff
@@ -16,7 +16,14 @@ import (
+	UnExposePorts() error
+}
+
+type VSock struct{}
```
Maybe it could be `virtualMachine` which implements the `AbstractVsock` interface instead of introducing a dummy/empty type for it? (just some thoughts to discuss, I'm not telling you this change must be done :)

rohanKanojia replied:
You're right about this. It's a cleaner approach than what I had added. Thanks a lot for suggesting this!

rohanKanojia committed:
Add VirtualMachine interface to abstract out crc virtual machine interaction

Add a VirtualMachine interface and make the CRC `machine` package client use the VirtualMachine interface instead of a concrete implementation. This way we can inject a dummy test FakeVirtualMachine implementation into client tests that can ease writing tests for this package.
  - Add some additional methods in VirtualMachine interface so that we can replace direct usage of struct fields with interface methods
    - `Bundle()`
    - `Driver()`
    - `API()`
    - `Host()`
    - `Kill()`

rohanKanojia committed:
test : Add a dummy VirtualMachine implementation for use in tests

Add FakeVirtualMachine sturct in `fakemachine` for adding dummy
implementation for `VirtualMachine` interface. Currently, I've
only completed methods used by `stop_test.go`, I'll add more in
small increments as we implement more unit tests using this
implementation.

Signed-off-by: Rohan Kumar <rohaan@redhat.com>

rohanKanojia committed:
test (machine) : Add additional tests in stop_test.go to use dummy VirtualMachine implementation

Add additional tests in `stop_test.go` to verify that client.Stop()
updates the state of virtual machine and unexposes exposed ports
as expected. Use the fake vm implementation added previously to
test vm state modification behavior on stop.

Signed-off-by: Rohan Kumar <rohaan@redhat.com>

cfergeau committed:
allow to provide fake vm implementation

Introduce a new constructor method `newClientWithVirtualMachine` in machine
client that would have an additional VirtualMachine argument, this would be
kept package private so that it's used only by tests in the same package.

cfergeau commented on pkg/crc/machine/client.go:
```diff
@@ -43,7 +43,7 @@ func NewClient(name string, debug bool, config crcConfig.Storage) Client {
 }
 
-// newClientWithVirtualMachine creates a Client instance with a overridden VirtualMachine implementation.
+// newClientWithVirtualMachine creates a Client instance with an overridden VirtualMachine implementation.
```
You could squash this in the commit where it was added.

cfergeau commented on pkg/crc/machine/stop_test.go:
```diff
@@ -6,11 +6,71 @@ import (
+	assert.NoError(t, stopErr)
+	assert.Equal(t, clusterState, state.Stopped)
+	assert.Equal(t, virtualMachine.IsStopped, true)
+	assert.Equal(t, virtualMachine.FakeSSHClient.LastExecutedCommand, "sudo -- sh -c 'crictl stop $(crictl ps -q)'")
```
In my opinion, the fact that `crictl stop` is the last command we run in the VM should be an implementation detail, it can change at any time if  we refactor this code, if we add new things to the bundle which needs to be stopped, ...

rohanKanojia replied:
Thanks for review, I'll remove this assertion.

praveenkumar commented:
We might not using virtualmachine object for crc-ng where vm creation would be part of macadam.
